src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MaxLengthAttribute.cs

```diff
@@ -76,11 +79,16 @@ public override bool IsValid(object value)
             {
                 length = str.Length;
             }
+            else if (value is ICollection collection)
+            {
+                length = collection.Count;
+            }
             else
             {
-                if (value is ICollection collection)
+                Type genericCol = value.GetType().GetTypeInfo().ImplementedInterfaces.Select(t => t.GetTypeInfo()).FirstOrDefault(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(ICollection<>));
```

Comment:
What should be the behavior if `ICollection<T>` is implemented multiple times? Options I see are:
* Throw
* Choose one in an arbitrary but deterministic way
* Don't do it this way

Throwing isn't very helpful because there probably isn't anything that can be done in the application to make it work.

I'm not against choosing one implementation deterministically, but I'm not convinced that the above code is deterministic across platforms/architectures, since it relies on Reflection ordering. We probably should do something like ordering by the name of the generic type.

A different approach would be to just look for a "Count" property with an integer return type on the actual type of value.GetType--essentially pattern matching, rather than using the interface typing. This would be my preference.

Reply:
So should I change it to just look for a `Count` property?
BTW @ajcvickers did you see #23578?

Reply:
@weitzhandler That would be my preference. I saw the ICountable discussion; I have no strong feelings one way or another on that--that is, done carefully it probably won't cause harm (i.e. break anything) but I also don't see a huge amount of value.

Reply:
The real amount of value is IMHO that the `Count` property is exposed the wrong way in .NET collections.